Preparations to support extensible non helper SymbolReferenceTable enum

0xdaryl opened the pull request:
* Relocate OpenJ9-specific SymbolReference functions to OpenJ9
* Make NonHelperSymbols enum extensible

0xdaryl commented:
@jamesgua : commit 047586d4cede0aec530c02e403f8b0f2c8e72f10 demonstrates generally the kinds of changes that you'll have to make in your project.  There is a bit of a dance in getting the OMR piece (eclipse/omr#5818) and this PR merged, but this should give you an idea of what to expect when the PRs eventually land.

0xdaryl commented on runtime/compiler/compile/J9SymbolReferenceTable.cpp:
```diff
@@ -2563,3 +2564,64 @@ J9::SymbolReferenceTable::createParameterSymbol(
+const char *
+J9::SymbolReferenceTable::getNonHelperSymbolName(CommonNonhelperSymbol nonHelper)
+   {
+#ifdef OMR_ENABLE_NONHELPER_EXTENSIBLE_ENUM
```
Note to reviewers: once this PR is committed, this macro will be defined upstream in OMR.  Once that macro definition propagates to OpenJ9 builds then this line can be removed.

mstoodle replied:
you meant to say "once https://github.com/eclipse/omr/pull/5822 is merged", right?

0xdaryl replied:
uh, yes.   Not sure what I was thinking there.

mstoodle commented on runtime/compiler/compile/J9NonHelperSymbols.enum:
```diff
@@ -0,0 +1,70 @@
+   lowTenureAddressSymbol,    // on j9vmthread
+   highTenureAddressSymbol,   // on j9vmthread
+   fragmentParentSymbol,
+   globalFragmentSymbol,
```
aren't the 4 above defined in OMR ?

0xdaryl replied:
I don't think so...   The `lowTenureAddressSymbol` and `highTenureAddressSymbol` appear once within a `J9_PROJECT_SPECIFIC` block in LoopVersioner, but that's about it.   You're right that that entire project-specific chunk could be refactored at some point though.

0xdaryl replied:
I think there is a problem with classRomPtr enum.  I'll look into it shortly.  I'll keep this WIP in the meantime.

0xdaryl replied:
Sorry it took so long to get back to this.  I checked each entry in the enum and it is fine (note that the enum is not "enabled" as a result of this PR, just the J9-specific entries are added to it).  The `classRomPtr` generation function is also migrated (via duplication) into the J9 namespace.  It will be cleaned up in OMR 5822.

mstoodle replied:
sorry, I didn't mean the JIT code that accesses those fields. I thought the fields themselves (on j9vmthread or in the GC structure holding the fragment values) were defined in OMR...

mstoodle replied:
@0xdaryl aren't the corresponding fields (`lowTenureAddress`, `highTenureAddress`, `fragmentParent`, and `globalFragment`) defined in OMR? In which case it seems a bit strange to be moving them to OpenJ9, unless you think those fields should also be moved to OpenJ9 (at some point?) ...

0xdaryl replied:
Yes, I understand what you mean now and I removed those 4 fields from the list.  I will add them back to OMR in the follow-on OMR PR.

I think ultimately they may move to OpenJ9, but as they represent fields in data structures that are in OMR and not in the compiler there will be effort required in considering their purpose is and migrating to OpenJ9.  I'll save that for another day.

mstoodle reviewed (commented):
Looks all right to me, just wondering if 4 of the symbols you're moving to J9 may need more refactoring work in OMR than you're anticipating...

0xdaryl commented:
I'm presently waiting for the OMR Acceptance to complete successfully before un-WIPing this.

0xdaryl commented:
Jenkins test sanity all jdk8,jdk11

0xdaryl commented:
Jenkins test sanity xlinux jdk8

0xdaryl commented:
Jenkins test sanity all jdk8,jdk11

0xdaryl commented:
I've un-WIPed this now as I believe I have addressed all the issues.  @mstoodle , would you mind taking a second look when you have a few moments please?

0xdaryl commented:
Jenkins test sanity osx,xlinuxxl jdk8

mstoodle commented:
Jenkins test sanity win jdk8

0xdaryl commented:
Jenkins test sanity all jdk8,jdk11

0xdaryl commented:
Jenkins test sanity win jdk11

0xdaryl commented:
Jenkins test sanity win jdk11

Windows failures are infrastructural.  Trying again.

0xdaryl commented:
@mstoodle : all checks passed (finally).  Please review again when you have a moment.

0xdaryl commented:
@mstoodle : the world's meekest poke

mstoodle reviewed (approved):
I am not actually sure `heapTop` and `heapBase` are used anywhere, so you could consider just removing them from the J9 specific list. But fully removing them is likely another PR anyway, so could be done then too. I'll approve but won't merge until you indicate which way you'd prefer to do it; I have no strong preference.

0xdaryl commented:
Thanks @mstoodle.  I may as well remove as part of this PR.  I'll push a new commit sometime today.

0xdaryl commented:
Jenkins test sanity all jdk11
